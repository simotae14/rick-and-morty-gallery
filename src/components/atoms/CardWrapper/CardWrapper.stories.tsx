import type { Meta, StoryObj } from '@storybook/react';

import CardWrapper from './index';

const meta = {
  title: 'Atoms/CardWrapper',
  component: CardWrapper,
  tags: ['autodocs'],
  argTypes: {
    cardWrapperType: {
      control: {
        options: ['base', 'spinner', 'card'],
        type: 'select',
      },
    },
  },
} satisfies Meta<typeof CardWrapper>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    cardWrapperType: undefined,
  },
};

export const Base: Story = {
  args: {
    cardWrapperType: 'base',
  },
};

export const Card: Story = {
  args: {
    cardWrapperType: 'card',
  },
};

export const Spinner: Story = {
  args: {
    cardWrapperType: 'spinner',
  },
};
