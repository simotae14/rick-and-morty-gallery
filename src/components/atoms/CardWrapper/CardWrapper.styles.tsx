import styled from 'styled-components';

export const DefaultWrapper = styled.div`
  background-color: #ffffff;
  box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);
  border-radius: 16px;
  -webkit-transition: all 0.2s ease-in;
  -moz-transition: all 0.2s ease-in;
  -ms-transition: all 0.2s ease-in;
  -o-transition: all 0.2s ease-in;
  transition: all 0.2s ease-in;
  &:hover {
    -webkit-transition: all 0.2s ease-in;
    -moz-transition: all 0.2s ease-in;
    -ms-transition: all 0.2s ease-in;
    -o-transition: all 0.2s ease-in;
    transition: all 0.2s ease-in;
    box-shadow: 0px 8px 12px rgba(0, 0, 0, 0.2);
  }
`;

export const BaseWrapper = styled(DefaultWrapper)`
  max-width: 408px;
  padding: 32px;
  position: relative;
  height: fit-content;
  width: fill-available;
  width: -webkit-fill-available;
`;

export const SpinnerWrapper = styled(DefaultWrapper)`
  width: 240px;
  height: 240px;
  margin: 112px auto;
`;

export const RickCardWrapper = styled(DefaultWrapper)`
  position: relative;
  margin: 16px 16px;
  width: 100%;
  max-width: 272px;
  height: 208px;
`;

export const RickCardSelectedWrapper = styled(RickCardWrapper)`
  border: 2px solid #121c33;
`;

export const DetailCardWarapper = styled(BaseWrapper)`
  margin: 36px 0 0 0;
`;
