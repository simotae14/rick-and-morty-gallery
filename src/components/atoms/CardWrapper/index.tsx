import {
  BaseWrapper,
  SpinnerWrapper,
  RickCardWrapper,
  RickCardSelectedWrapper,
  DetailCardWarapper,
} from './CardWrapper.styles.jsx';

export enum CardWrapperEnum {
  base = 'base',
  spinner = 'spinner',
  card = 'card',
  selected = 'selected',
  detail = 'detail',
}

type CardWrapperType = `${CardWrapperEnum}`;

const getCardWrapper = (cardWrapperType: CardWrapperType) => {
  return {
    [CardWrapperEnum.base]: BaseWrapper,
    [CardWrapperEnum.spinner]: SpinnerWrapper,
    [CardWrapperEnum.card]: RickCardWrapper,
    [CardWrapperEnum.selected]: RickCardSelectedWrapper,
    [CardWrapperEnum.detail]: DetailCardWarapper,
  }[cardWrapperType];
};

/**
 * Component that Wraps the card
 */
function CardWrapper({
  cardWrapperType = CardWrapperEnum.base,
  onClickHandler,
  children,
}: React.PropsWithChildren<{
  /**
   * Is this the possible variant of the CardWrapper
   */
  cardWrapperType?: CardWrapperType;
  /**
   * Is the action triggered on clicking the CardWrapper
   */
  onClickHandler?: () => void;
}>) {
  const CustomCardWrapper = getCardWrapper(cardWrapperType);

  return (
    <CustomCardWrapper onClick={onClickHandler}>{children}</CustomCardWrapper>
  );
}

export default CardWrapper;
