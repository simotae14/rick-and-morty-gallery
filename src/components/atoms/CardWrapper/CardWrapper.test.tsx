/* eslint-disable import/no-extraneous-dependencies */
import { describe, it } from 'vitest';
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';

import CardWrapper from './index';

expect.extend(toHaveNoViolations);

describe('Components => atoms => CardWrapper', () => {
  it('should render the base CardWrapper when no props are passed', () => {
    const { asFragment } = render(<CardWrapper>content</CardWrapper>);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should render the spinner CardWrapper when is passed the cardWrapperType "spinner"', () => {
    const { asFragment } = render(
      <CardWrapper cardWrapperType="spinner">content</CardWrapper>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('should render the spinner CardWrapper when is passed the cardWrapperType "card"', () => {
    const { asFragment } = render(
      <CardWrapper cardWrapperType="card">content</CardWrapper>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('should not fail any accessibility tests the base Card Wrapper', async () => {
    const { container } = render(<CardWrapper>content</CardWrapper>);
    expect(await axe(container)).toHaveNoViolations();
  });
  it('should not fail any accessibility tests the spinner Card Wrapper', async () => {
    const { container } = render(
      <CardWrapper cardWrapperType="spinner">content</CardWrapper>
    );
    expect(await axe(container)).toHaveNoViolations();
  });
  it('should not fail any accessibility tests the card Card Wrapper', async () => {
    const { container } = render(
      <CardWrapper cardWrapperType="card">content</CardWrapper>
    );
    expect(await axe(container)).toHaveNoViolations();
  });
});
