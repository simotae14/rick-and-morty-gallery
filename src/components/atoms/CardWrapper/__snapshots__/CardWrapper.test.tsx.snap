// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Components => atoms => CardWrapper > should render the base CardWrapper when no props are passed 1`] = `
<DocumentFragment>
  <div
    class="sc-eLqIir sc-HzRMQ kfzyXq hDBcnb"
  >
    content
  </div>
</DocumentFragment>
`;

exports[`Components => atoms => CardWrapper > should render the spinner CardWrapper when is passed the cardWrapperType "card" 1`] = `
<DocumentFragment>
  <div
    class="sc-eLqIir sc-buuUZy kfzyXq dcAPxy"
  >
    content
  </div>
</DocumentFragment>
`;

exports[`Components => atoms => CardWrapper > should render the spinner CardWrapper when is passed the cardWrapperType "spinner" 1`] = `
<DocumentFragment>
  <div
    class="sc-eLqIir sc-hIFLYB kfzyXq iDmrJc"
  >
    content
  </div>
</DocumentFragment>
`;
